controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -0,0 +1,663 @@
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileNormal(volCtx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VolumeContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachine for batch volume processing")
+	defer func() {
+		ctx.Logger.Info("Finished Reconciling VirtualMachine for batch volume processing")
+	}()
+
+	// Reconcile instance storage volumes
+	if pkgcfg.FromContext(ctx).Features.InstanceStorage {
+		ready, err := r.reconcileInstanceStoragePVCs(ctx)
+		if err != nil || !ready {
+			return err
+		}
+	}
+
+	if ctx.VM.Status.BiosUUID == "" {
+		// CNS requires the BiosUUID to match up the attachment request with the VM.
+		if len(ctx.VM.Spec.Volumes) != 0 {
+			ctx.Logger.Info("VM Status does not yet have BiosUUID. Deferring volume attachment")
+		}
+		return nil
+	}
+
+	// Add hardware version validation
+	if err := r.validateHardwareVersion(ctx); err != nil {
+		return fmt.Errorf("hardware version validation failed: %w", err)
+	}
+
+	// Get existing batch attachment for this VM
+	batchAttachment, err := r.getBatchAttachmentForVM(ctx)
+	if err != nil {
+		ctx.Logger.Error(err, "Error getting existing CnsNodeVmBatchAttachment for VM")
+		return err
+	}
+
+	// Handle existing batch attachment conflicts
+	if batchAttachment != nil {
+		if err := r.handleExistingBatchAttachment(ctx, batchAttachment); err != nil {
+			return fmt.Errorf("failed to handle existing batch attachment: %w", err)
+		}
+	}
+
+	// Process volumes and create/update batch attachment
+	if err := r.processBatchAttachment(ctx, batchAttachment); err != nil {
+		ctx.Logger.Error(err, "Error processing CnsNodeVmBatchAttachment")
+		return err
+	}
+
+	return nil
+}
+
+// getBatchAttachmentForVM returns the CnsNodeVmBatchAttachment
+// resource for the VM. We assume that the name of the resource
+// matches the name of the VM.
+func (r *Reconciler) getBatchAttachmentForVM(ctx *pkgctx.VolumeContext) (*cnsv1alpha1.CnsNodeVmBatchAttachment, error) {
+	attachment := &cnsv1alpha1.CnsNodeVmBatchAttachment{}
+
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name:      ctx.VM.Name,
+		Namespace: ctx.VM.Namespace,
+	}, attachment); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		return nil, nil
+	}
+
+	return attachment, nil
+}
+
+func (r *Reconciler) processBatchAttachment(
+	ctx *pkgctx.VolumeContext,
+	existingAttachment *cnsv1alpha1.CnsNodeVmBatchAttachment) error {
+
+	// Filter volumes that have PVC source
+	pvcVolumes := make([]vmopv1.VirtualMachineVolume, 0)
+	for _, vol := range ctx.VM.Spec.Volumes {
+		if vol.PersistentVolumeClaim != nil {
+			pvcVolumes = append(pvcVolumes, vol)
+		}
+	}
+
+	// Handle WFFC for each PVC volume
+	for _, vol := range pvcVolumes {
+		if err := r.handlePVCWithWFFC(ctx, vol); err != nil {
+			ctx.Logger.Error(err, "Failed to handle WFFC for volume", "volume", vol.Name)
+			// Continue processing other volumes
```

Comment:
Then we should keep track of error for each vol, so that later when we create BatchAttechment spec, we should skip adding these if there are error

Reply:
MM, I think this func could be put inside of `processVolumesWithErrorHandling `?